### PR TITLE
register fonts for use by cssbox

### LINKED ISSUE #11401: Slack Pulses aren't rendering Lato font correctly
This is not the typeface we've all come to know and love. This pulse was generated from our Stats instance.

image

### PR DESCRIPTION
**Pull request by flyingmachine**
fixes #11401

### CODE REVIEW

**Comment by flamber** [comment]
@flyingmachine Awesome you figured this out! Will it fail the build if the fonts are changed/missing?

**Comment by flyingmachine** [comment]
This one's tricky because it's not easily reproducible or testable. To reproduce it I had to run my local source code in a docker container to produce an environment where the operating system didn't provide the Lato font, and then send a pulse to Slack to check it out.

I've also created a new namespace, `dev.render-png` which has a function to make it easier to view rendered pngs from the REPL. However, pngs rendered from a REPL started on a developer's personal machine will not necessarily reflect the pngs rendered from a docker image.

**Comment by flyingmachine** [comment]
@flamber - it doesn't, good call!

**Comment by flyingmachine** [comment]
@flamber actually now that I've checked, the build will in fact fail :)

File: src/metabase/pulse/render/png.clj [hunk 1]
```diff
@@ -23,6 +23,21 @@
            org.fit.cssbox.layout.BrowserCanvas
            org.w3c.dom.Document))
 
+
+(defonce
+  ^{:doc     "Makes custom fonts available to Java so that CSSBox can render them"
+    :private true}
+  register-fonts
+  (delay (doseq [weight ["regular" "700" "900"]]
```
Comment (walterl) [inline src/metabase/pulse/render/png.clj]: You just _know_ that, if the set of fonts ever change, that this vector won't be updated.

Isn't it better, then, to do something like this?

```clojure
(->> (clojure.java.io/file "resources/frontend_client/app/fonts/lato-v16-latin")
     file-seq
     (filter #(re-find #"^lato-v16-latin-\w+\.ttf$" (.getName %))))
```

But this suffers from you-can't-list-files-in-a-jar's-resources syndrome, right?

**Review by walterl — COMMENTED** [review]
Do we need a tests for this? Are there any other resources that could face a similar situation?

**Comment by flyingmachine** [comment]
@walterl tests would be nice, but I honestly don't know how we would test it. My [earlier comment](https://github.com/metabase/metabase/pull/11577#issuecomment-568039589) hopefully explains why it's difficult to get to a point that this can get reproduced, and even if we are able to set up a good test environment I'm not really sure what we would test. Maybe compare a generated image against a canonical test image?

**Comment by walterl** [comment]
@flyingmachine How about a more indirect test, like testing that all of the resource fonts are registered in the app?

**Comment by flyingmachine** [comment]
@walterl I guess we could, but if any of the fonts are missing then an exception is thrown when the namespace is compiled, so missing fonts will already fail the build

**Comment by flyingmachine** [comment]
my preference is to not add such a test

File: src/metabase/pulse/render/png.clj (hunk 1 above)
Comment (flyingmachine) [reply src/metabase/pulse/render/png.clj]: yeah, the code won't pick up new fonts :( however i think for now i'd like to go ahead and merge the PR anyway. we could investigate some scheme for discovering and loading fonts, but i don't think that's a high priority

**Review by walterl — APPROVED** [review]
:+1:

**Pull request merged**